Modify variable product sync hooks

thenbrent opened the pull request:
This PR introduces two changes to variable product sync hooks:

1. The `'woocommerce_variable_product_sync'` hook is now triggered after a variable product is saved
2. A new hook is triggered before the variable product is saved

Justification for each change below.

### Trigger `'woocommerce_variable_product_sync'` Hook After Save

Because only a product ID is passed to callbacks on `'woocommerce_variable_product_sync'`, callbacks will need to instantiate their own instance of the product to do anything useful with it. This creates a lot of potential for problems given that the product has at that stage had its data only partially saved.

For example, a callback may modify some meta data on its own instance of the product, then save that, only to immediately have that overridden if the instance of `$product` in `WC_Product_Variable::sync()` also had changes to that meta data.

To avoid that, `WC_Product_Variable::sync()` can instead trigger `'woocommerce_variable_product_sync'` only after the product has been saved. This is backward compatible because the hook was triggered at the very end of the process in WC < 2.7.

### New `'woocommerce_variable_product_nsync'` Hook

Because the existing `'woocommerce_variable_product_sync'` hook only passes callbacks the product ID, there is no way for callbacks to modify the actual instance of the product being sync'd/saved by `WC_Product_Variable::sync()`. This matters a lot more in WC 2.7+ where product properties are abstracted from their storage.

For example, to set meta data on the product:
* in older versions of WC, it was possible to simply save meta data directly using the product ID with something like `update_post_meta( $product_id, $meta_key, $meta_value )`
* in WC 2.7, `$product->add_meta_data()` should be used to make sure it is saved in that's products data store, and also so that meta data can be saved in one go rather than multiple writes

The new `'woocommerce_variable_product_nsync'` hook passes callbacks the current instance of the variable product, so that it can be modified during synchronisation (potentially before being saved).

As far as the hook name goes, I couldn't help but try to sneak in a bad 90s boy band reference, especially when the only alternatives I could come up with were clumsy sounding hooks like `'woocommerce_variable_product_in_sync'` or `'woocommerce_variable_product_sync_before_save'`.

thenbrent committed:
Trigger variable product sync hook after save

Because only a product ID is passed to callbacks on 'woocommerce_variable_product_sync',
they will need to instantiate their own instance of the product to do anything useful
with it. This creates a lot of potential for problems given that the product has at that
stage had its data sync'd but not all of it saved. For example, a callback may modify some
meta data on its own instance of the product, then save that, only to immediately have
that overridden if the instance of $product in WC_Product_Variable::sync() also had changes
to that meta data.

Instead, trigger 'woocommerce_variable_product_sync' only after the product has been
saved. This is backward compatible because the hook was triggered at the very end of
the process in WC < 2.7.

thenbrent committed:
Remoe unused 'edit' param

To avoid a merge conflict with #13063

thenbrent committed:
Add new variable product sync hook

To provide a way for callbacks to modify the actual instance of the
variable product being sync'd, especially before its saved, in case
it needs to sync some meta data or other prop that can be set on the
object itself.

mikejolley commented:
@thenbrent I'm not fond of nsync :p 

You're right about action position, although I think we should go further and do a deprecation action hook to discourage usage.

How about also only passing the object, not `$save` as thats irrelevant to syncing functions?

thenbrent commented:
> @thenbrent I'm not fond of nsync :p

😭 

> I think we should go further and do a deprecation action hook to discourage usage.

👍 

> How about also only passing the object, not $save as thats irrelevant to syncing functions?

I'll submit a revised patch this afternoon taking care of those 3 things. 😄

thenbrent committed:
Merge branch 'master' into variable_product_sync_filters

Conflicts:
	includes/class-wc-product-variable.php

thenbrent committed:
Deprecate 'woocommerce_variable_product_sync'

thenbrent committed:
Add woocommerce_variable_product_sync_before_save

Removing woocommerce_variable_product_nsync filter and $save param.

thenbrent committed:
Add 'woocommerce_variable_product_synced' filter

thenbrent commented:
@mikejolley just pushed up patches addressing https://github.com/woocommerce/woocommerce/pull/13064#issuecomment-278059756 I've also merged in master to address the merge conflict because of #13063.

The patch now:

* deprecates the `'woocommerce_variable_product_sync'` filter
* includes two new filters, one before saving and one after
* doesn't pass `$save` to callbacks on those filters

mikejolley commented on includes/class-wc-product-variable.php:
```diff
@@ -455,11 +455,15 @@ public static function sync( $product, $save = true ) {
 			self::sync_attributes( $product ); // Legacy update of attributes.
 
-			do_action( 'woocommerce_variable_product_sync', $product->get_id(), $product->get_visible_children(), $save );
+			do_action( 'woocommerce_variable_product_sync_before_save', $product );
```
Save is also an 'optional' thing. TBH I think a hook name such as `woocommerce_variable_product_sync_data` would be fine. We'd have used `woocommerce_variable_product_sync` if available but it wouldn't be backwards compatible.

mikejolley commented on includes/class-wc-product-variable.php:
```diff
@@ -455,11 +455,15 @@ public static function sync( $product, $save = true ) {
 				$product->save();
 			}
+
+			do_action( 'woocommerce_variable_product_synced', $product );
```
I don't agree that this one is needed. Unless `$save` is true, this hook is identical to the one above, and if you really did need to kick logic in at this point there are action hooks in each object's save/create/update methods which can be used instead so you can capture all save events.

Did I miss a use case?

thenbrent replied:
The main use case I had in mind was an easy replacement for `'woocommerce_variable_product_sync'`.

I checked for suitable replacements in `WC_Product_Variable::save()`, but there was only the `'woocommerce_before_' . $this->object_type . '_object_save'` hook, which runs before saving.

It looks like further upstream `'woocommerce_new_product'` or `'woocommerce_new_product'` would be suitable though, so I'll remove this and mention those in the deprecated notice for `'woocommerce_variable_product_sync'`.

mikejolley commented on includes/class-wc-product-variable.php:
```diff
@@ -455,11 +455,15 @@ public static function sync( $product, $save = true ) {
+
+			do_action( 'woocommerce_variable_product_synced', $product );
+
+			wc_do_deprecated_action( 'woocommerce_variable_product_sync', array( $product->get_id(), $product->get_visible_children() ), '2.7', 'woocommerce_variable_product_sync_before_save' );
```
👍

mikejolley reviewed (changes requested):
Just a few notes.

thenbrent committed:
Add 'woocommerce_variable_product_sync_data'

Renaming 'woocommerce_variable_product_sync_before_save'

thenbrent committed:
Remove 'woocommerce_variable_product_synced' hook

And add additional suggested alternatives to the deprecated
woocommerce_variable_product_sync hook.

thenbrent commented:
@mikejolley I've addressed those comments. For an explanation on why I've also added `'woocommerce_new_product'` & `'woocommerce_update_product'` to the deprecated notice in SHA: 5d560f3 see https://github.com/woocommerce/woocommerce/pull/13064#discussion_r100595362
